cli/internal/cloudcmd/upgrade.go

```diff
@@ -74,100 +76,150 @@ func NewUpgrader(outWriter io.Writer, log debugLog) (*Upgrader, error) {
 		stableInterface:  &stableClient{client: kubeClient},
 		dynamicInterface: &dynamicClient{client: unstructuredClient},
 		helmClient:       helmClient,
-		outWriter:        outWriter,
-		log:              log,
+
+		outWriter: outWriter,
+		log:       log,
 	}, nil
 }
 
-// UpgradeImage upgrades the cluster to the given measurements and image.
-func (u *Upgrader) UpgradeImage(ctx context.Context, newImageReference, newImageVersion string, newMeasurements measurements.M) error {
-	nodeVersion, err := u.getConstellationVersion(ctx)
-	if err != nil {
-		return fmt.Errorf("retrieving current image: %w", err)
+// UpgradeHelmServices upgrade helm services.
+func (u *Upgrader) UpgradeHelmServices(ctx context.Context, config *config.Config, timeout time.Duration, allowDestructive bool) (string, error) {
+	err := u.helmClient.Upgrade(ctx, config, timeout, allowDestructive)
+	invalidUpgrade := &compatibility.InvalidUpgradeError{}
+	if errors.As(err, &invalidUpgrade) {
+		return fmt.Sprintf("Skipping Microservice upgrade: %s", err), nil
 	}
-	currentImageVersion := nodeVersion.Spec.ImageVersion
+	return "", err
+}
 
-	if err := compatibility.IsValidUpgrade(currentImageVersion, newImageVersion); err != nil {
-		return err
+// UpgradeNodeVersion upgrades the cluster's NodeVersion object and in turn triggers image & k8s version upgrades.
+// The versions set in the config are validated against the versions running in the cluster.
+func (u *Upgrader) UpgradeNodeVersion(ctx context.Context, imageFetcher ImageFetcher, conf *config.Config) (string, error) {
```

Comment:
Same thing here.
I think this function would be cleaner if it just returned errors, or handled printing the warnings/errors itself